File: lib/services/offline_media_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:hive_flutter/hive_flutter.dart';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';
import 'package:path/path.dart' as path;

import '../models/media_item.dart';

/// Stores media list and files locally via Hive + file system for offline playback.
/// After restart with no internet, media is loaded from Hive and played from cached files.
class OfflineMediaService {
  static const _boxName = 'display_cache';
  static const _listPrefix = 'list_';
  static const _pathsPrefix = 'paths_';
  static const _mediaDirName = 'cms_media';

  Box<dynamic>? _box;

  Future<void> init() async {
    if (_box != null) return;
    await Hive.initFlutter();
    _box = await Hive.openBox(_boxName);
  }

  String _listKey(String mac) => '$_listPrefix$mac';
  String _pathsKey(String mac) => '$_pathsPrefix$mac';

  /// Save media list for this MAC (e.g. after successful API fetch).
  Future<void> saveMediaList(String mac, List<MediaItem> list) async {
    await init();
    final maps = list.map((m) => _itemWithLocalPath(m, mac)).toList();
    await _box!.put(_listKey(mac), jsonEncode(maps));
  }

  Map<String, dynamic> _itemWithLocalPath(MediaItem m, String mac) {
    final map = m.toMap();
    final path = getLocalPath(mac, m.id);
    if (path != null) map['local_path'] = path;
    return map;
  }

  /// Load media list from local storage. Returns null if none saved.
  Future<List<MediaItem>?> getMediaList(String mac) async {
    await init();
    final data = _box!.get(_listKey(mac));
    if (data == null || data is! String) return null;
    List<dynamic> decoded;
    try {
      decoded = jsonDecode(data) as List<dynamic>;
    } catch (_) {
      return null;
    }
    final paths = _getPathsMap(mac);
    final list = <MediaItem>[];
    for (final e in decoded) {
      if (e is Map<String, dynamic>) {
        final map = Map<String, dynamic>.from(e);
        final id = (map['id'] as num?)?.toInt();
        if (id != null && paths[id.toString()] != null) {
          map['local_path'] = paths[id.toString()];
        }
        list.add(MediaItem.fromMap(map));
      }
    }
    return list.isEmpty ? null : list;
  }

  Map<String, String> _getPathsMap(String mac) {
    final data = _box!.get(_pathsKey(mac));
    if (data == null || data is! String) return {};
    try {
      final decoded = jsonDecode(data) as Map<String, dynamic>;
      return decoded.map((k, v) => MapEntry(k, v?.toString() ?? ''));
    } catch (_) {
      return {};
    }
  }

  /// Get local file path for a media id if cached.
  String? getLocalPath(String mac, int mediaId) {
    if (_box == null) return null;
    final paths = _getPathsMap(mac);
    return paths[mediaId.toString()];
  }

  /// Save local path for a media id (after downloading).
  Future<void> saveLocalPath(String mac, int mediaId, String filePath) async {
    await init();
    final paths = _getPathsMap(mac);
    paths[mediaId.toString()] = filePath;
    await _box!.put(_pathsKey(mac), jsonEncode(paths));
  }

  /// Directory for cached media files for this MAC.
  Future<Directory> _mediaDir(String mac) async {
    final root = await getApplicationDocumentsDirectory();
    final dir = Directory(path.join(root.path, _mediaDirName, _sanitizeMac(mac)));
    if (!await dir.exists()) await dir.create(recursive: true);
    return dir;
  }

  String _sanitizeMac(String mac) => mac.replaceAll(RegExp(r'[^a-zA-Z0-9]'), '_');

  String _extensionFromMime(String fileType) {
    if (fileType.contains('jpeg') || fileType.contains('jpg')) return 'jpg';
    if (fileType.contains('png')) return 'png';
    if (fileType.contains('gif')) return 'gif';
    if (fileType.contains('webp')) return 'webp';
    if (fileType.contains('mp4')) return 'mp4';
    if (fileType.contains('webm')) return 'webm';
    if (fileType.contains('mov')) return 'mov';
    return 'bin';
  }

  /// Download media file and save locally; store path in Hive.
  Future<String?> cacheMediaFile(String mac, MediaItem media) async {
    try {
      final dir = await _mediaDir(mac);
      final ext = _extensionFromMime(media.fileType);
      final file = File(path.join(dir.path, '${media.id}.$ext'));
      final response = await http.get(Uri.parse(media.previewUrl));
      if (response.statusCode != 200) return null;
      await file.writeAsBytes(response.bodyBytes);
      await saveLocalPath(mac, media.id, file.path);
      return file.path;
    } catch (_) {
      return null;
    }
  }

  /// Cache all media files in background. Call after saving list when online.
  Future<void> cacheAllInBackground(String mac, List<MediaItem> list) async {
    for (final media in list) {
      if (!media.isPlayable) continue;
      if (getLocalPath(mac, media.id) != null) continue;
      await cacheMediaFile(mac, media);
    }
    await _refreshListWithPaths(mac);
  }

  /// After caching files, update stored list so each item has local_path.
  Future<void> _refreshListWithPaths(String mac) async {
    final list = await getMediaList(mac);
    if (list == null) return;
    final updated = list.map((m) {
      final p = getLocalPath(mac, m.id);
      return MediaItem(
        id: m.id,
        title: m.title,
        fileType: m.fileType,
        fileSize: m.fileSize,
        previewUrl: m.previewUrl,
        updatedAt: m.updatedAt,
        localPath: p,
      );
    }).toList();
    await saveMediaList(mac, updated);
  }

  /// Merge API list with any existing local paths so playback uses cache when available.
  List<MediaItem> mergeWithLocalPaths(String mac, List<MediaItem> fromApi) {
    return fromApi.map((m) {
      final local = getLocalPath(mac, m.id);
      if (local == null) return m;
      return MediaItem(
        id: m.id,
        title: m.title,
        fileType: m.fileType,
        fileSize: m.fileSize,
        previewUrl: m.previewUrl,
        updatedAt: m.updatedAt,
        localPath: local,
      );
    }).toList();
  }
}

File: cms_display_android/app/src/main/java/com/cms/display/offline/OfflineMediaService.kt
package com.cms.display.offline

import android.content.Context
import com.cms.display.model.MediaItem
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.net.URL
import java.util.regex.Pattern

class OfflineMediaService(private val context: Context) {

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val listPrefix = "list_"
    private val pathsPrefix = "paths_"
    private val mediaDirName = "cms_media"

    fun saveMediaList(mac: String, list: List<MediaItem>) {
        val arr = JSONArray()
        for (m in list) {
            val obj = JSONObject().apply {
                put("id", m.id)
                put("title", m.title)
                put("file_type", m.fileType)
                put("file_size", m.fileSize)
                put("preview_url", m.previewUrl)
                m.playbackScheduleSeconds?.let { put("playback_schedule_seconds", it) }
                put("updated_at", m.updatedAt)
                put("local_path", getLocalPath(mac, m.id) ?: m.localPath)
            }
            arr.put(obj)
        }
        prefs.edit().putString(listKey(mac), arr.toString()).apply()
    }

    fun getMediaList(mac: String): List<MediaItem>? {
        val json = prefs.getString(listKey(mac), null) ?: return null
        return try {
            val arr = JSONArray(json)
            val paths = getPathsMap(mac)
            (0 until arr.length()).mapNotNull { i ->
                val obj = arr.optJSONObject(i) ?: return@mapNotNull null
                val id = obj.optInt("id")
                val localPath = paths[id.toString()] ?: obj.optString("local_path").takeIf { it.isNotEmpty() }
                val scheduleSeconds: Long? = if (obj.has("playback_schedule_seconds") && !obj.isNull("playback_schedule_seconds")) {
                    obj.optLong("playback_schedule_seconds", 0L).takeIf { it > 0L }
                } else null
                MediaItem(
                    id = id,
                    title = obj.optString("title").takeIf { it.isNotEmpty() },
                    fileType = obj.optString("file_type", "image/jpeg"),
                    fileSize = obj.optInt("file_size", 0),
                    previewUrl = obj.optString("preview_url"),
                    playbackScheduleSeconds = scheduleSeconds,
                    updatedAt = obj.optString("updated_at").takeIf { it.isNotEmpty() },
                    localPath = localPath
                )
            }.takeIf { it.isNotEmpty() }
        } catch (_: Exception) {
            null
        }
    }

    fun getLocalPath(mac: String, mediaId: Int): String? =
        getPathsMap(mac)[mediaId.toString()]

    fun saveLocalPath(mac: String, mediaId: Int, filePath: String) {
        val paths = getPathsMap(mac).toMutableMap()
        paths[mediaId.toString()] = filePath
        val obj = JSONObject()
        paths.forEach { (k, v) -> obj.put(k, v) }
        prefs.edit().putString(pathsKey(mac), obj.toString()).apply()
    }

    fun mergeWithLocalPaths(mac: String, fromApi: List<MediaItem>): List<MediaItem> {
        val paths = getPathsMap(mac)
        return fromApi.map { m ->
            paths[m.id.toString()]?.let { path -> m.withLocalPath(path) } ?: m
        }
    }

    fun cacheAllInBackground(mac: String, list: List<MediaItem>) {
        Thread {
            for (media in list) {
                if (!media.isPlayable || getLocalPath(mac, media.id) != null) continue
                cacheMediaFile(mac, media)
            }
            refreshListWithPaths(mac)
        }.start()
    }

    fun cacheMediaFile(mac: String, media: MediaItem): String? {
        return try {
            val dir = mediaDir(mac)
            val ext = extensionFromMime(media.fileType)
            val file = File(dir, "${media.id}.$ext")
            URL(media.safePreviewUrl).openStream().use { input ->
                file.outputStream().use { output ->
                    input.copyTo(output)
                }
            }
            saveLocalPath(mac, media.id, file.absolutePath)
            file.absolutePath
        } catch (_: Exception) {
            null
        }
    }

    private fun listKey(mac: String) = listPrefix + mac
    private fun pathsKey(mac: String) = pathsPrefix + mac

    private fun getPathsMap(mac: String): Map<String, String> {
        val json = prefs.getString(pathsKey(mac), null) ?: return emptyMap()
        return try {
            val obj = JSONObject(json)
            obj.keys().asSequence().associateWith { obj.optString(it, "") }
        } catch (_: Exception) {
            emptyMap()
        }
    }

    private fun mediaDir(mac: String): File {
        val dir = File(context.filesDir, "$mediaDirName/${sanitizeMac(mac)}")
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    private fun sanitizeMac(mac: String) = mac.replace(Regex("[^a-zA-Z0-9]"), "_")

    private fun extensionFromMime(fileType: String): String {
        return when {
            fileType.contains("jpeg") || fileType.contains("jpg") -> "jpg"
            fileType.contains("png") -> "png"
            fileType.contains("gif") -> "gif"
            fileType.contains("webp") -> "webp"
            fileType.contains("mp4") -> "mp4"
            fileType.contains("webm") -> "webm"
            fileType.contains("mov") -> "mov"
            else -> "bin"
        }
    }

    private fun refreshListWithPaths(mac: String) {
        val list = getMediaList(mac) ?: return
        val updated = list.map { m ->
            getLocalPath(mac, m.id)?.let { m.withLocalPath(it) } ?: m
        }
        saveMediaList(mac, updated)
    }

    companion object {
        private const val PREFS_NAME = "display_cache"
    }
}
